docs/chapters/architecture/DLT/diagnostics-log-and-tracing.rst

```diff
@@ -0,0 +1,177 @@
+
+Stakeholders
+^^^^^^^^^^^^
+* Software Developers
+* QA Engineers
+
+Constraints
+-----------
+GENIVI's dlt-daemon stores an offline log database on in-RAM file system /tmp/dlt_offlinetrace_20120101_005302.dlt . The size limits for the log files should be adjusted at HU startup based on amount of RAM available. E.g. a Premium HU with 16GB of RAM can use 200MB of offline storage. On the other hand, 4GB of ram uses 50MB of offline storage or less.
+
+Context
+-------
+Business Context
+^^^^^^^^^^^^^^^^
+DLT logs are automotive industry standard. Industry expects any automotive software to support it.
+
+Solution Strategy
+-----------------
+
+Since PELUX is a part of open source community, we use a component developed by GENIVI.
+
+Below are the open source software that we use from the DLT component.
+
+* `GENIVI dlt-daemon`_
+* `GENIVI dlt-viewer`_
+
+Since GENIVI dlt-daemon supports only C applications, another third party software is required to support C++. Currently, PELUX uses `ivi-logging`_. However, wrapper discussions are still ongoing. We currently have three options which are compared under Design Decisions.
+
+Building blocks view
+--------------------
+
+Level 1
+^^^^^^^
+.. image:: graphs/block-view-lvl1.png
+
+Level 2 
+^^^^^^^
+.. image:: graphs/block-view-lvl2.png
+
+Runtime view
+------------
+Depending on what the developers decide to log and trace, the logs can be viewed via dlt-viewer.
+
+Optionally, Blue PiraT 2 can be configured to work with dlt-daemon, through which multiple gigabytes of logs can be stored.
+
+
+Design decisions
+----------------
+
+C++ wrapper
+^^^^^^^^^^^
+
+`ivi-logging`_
+""""""""""""""
+This component was initially developed by Pelagicore but is not actively maintained at this time.
+
+Pros:
+
+* It exposes all functions of libdlt.so library.
+* C++11
+
+Cons:
+
+* It is a header-based library. Each translation unit (\*.o) will have around 20Kb of code segment overhead. Each binary file will be bloated and heavily flooded with inline code.
+* Increases compilation time of every translation unit where it is used.
+* Nobody is familiar with it
+
+`log4cplus`_
+""""""""""""
+A popular logging library that inherits log4java principles and log formats
+
+Pros:
+
+* It is very popular. Many c++ developers already familiar with it.
+* Well maintained.
+* Do not have dependencies.
+* Can write to many sinks (console, socket, syslog)
+* DLT sink can be easily added
+* C++11
+
+Cons:
+
+* A full set of DLT features will NOT be available to app developer.
+* All primitive types (int32, float, double) will be converted to string logs.
+* LogLevel setup from DLTAppender side is not easy.  LogLevel is set on Logger level, not on Appender level.
+* DLT AppID and ContexID registration will require additional functions that breaks standard API
+
+`CAmDltWrapper (GenIVI AudioManager)`_
+""""""""""""""""""""""""""""""""""""""
+GenIVI AudioManager is a C++ program and it has it's own wrapper around DLT logging system.
```

Comment:
Simimlarly to what I said about PELUX/Pelux/pelux. I'm confused by this "GenIVI" now.

Reply:
it has _its_ own